bundle/src/main/java/com/adobe/acs/commons/ondeploy/impl/OnDeployExecutorImpl.java

```diff
@@ -0,0 +1,273 @@
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.commons.classloader.DynamicClassLoaderManager;
+import org.apache.sling.commons.osgi.PropertiesUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.Value;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A service that triggers scripts on deployment to an AEM server.
+ * <p>
+ * This class manages scripts so that they only run once (unless the script
+ * fails).  Script execution statuses are stored in the JCR @
+ * /var/acs-commons/on-deploy-scripts-status.
+ * <p>
+ * Scripts are specified via OSGi config, are are run in the order specified.
+ * <p>
+ * NOTE: Since it's always a possibility that
+ * /var/acs-commons/on-deploy-scripts-status will be deleted in the JCR,
+ * scripts should be written defensively in case they are actually run more
+ * than once.  This also covers the scenario where a script is run a second
+ * time after failing the first time.
+ */
+@Component(
+        label = "ACS AEM Commons - On-Deploy Scripts Executor",
+        description = "Developer tool that triggers scripts (specified via an implementation of OnDeployScriptProvider) to execute on deployment.",
+        immediate = true,
+        metatype = true,
+        policy = ConfigurationPolicy.REQUIRE,
+        configurationFactory = true
+)
+@Service
+public class OnDeployExecutorImpl implements OnDeployExecutor {
+    @Property(
+            name = "scripts",
+            cardinality = Integer.MAX_VALUE,
+            label = "Scripts",
+            description = "Classes that repersent on-deploy scripts (implement OnDeployScript), in the order to run them."
+    )
+    private static final String PROP_SCRIPTS = "scripts";
+
+    private static final String SCRIPT_DATE_END = "endDate";
+    private static final String SCRIPT_DATE_START = "startDate";
+    private static final String SCRIPT_STATUS = "status";
+    private static final String SCRIPT_STATUS_JCR_FOLDER = "/var/acs-commons/on-deploy-scripts-status";
+    private static final String SCRIPT_STATUS_FAIL = "fail";
+    private static final String SCRIPT_STATUS_RUNNING = "running";
+    private static final String SCRIPT_STATUS_SUCCESS = "success";
+
+    private static final String SERVICE_NAME = "on-deploy-scripts";
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @Reference
+    private QueryBuilder queryBuilder;
+    @Reference
+    private ResourceResolverFactory resourceResolverFactory;
+    @Reference
+    private DynamicClassLoaderManager dynamicClassLoaderManager;
+
+    private List<OnDeployScript> scripts = null;
+
+    // TODO: These were needed for injection to work in tests
```

Comment:
@justinedelson @davidjgonzalez My sling-mock-fu is weak, and for whatever reason I was required to have these `bind` functions in order for the `@Reference` fields above to be injected correctly.

Is there a way around that, or can either of you provide insight as to why these are required and/or if its a good thing to keep these?